src/Utils/Rotation.cpp

```diff
@@ -312,15 +312,335 @@ void utils::Rotation::checkUnitary()
+            det = Mk(0, 0) * M_adj_Tk(0, 0) + Mk(0, 1) * M_adj_Tk(0, 1) + Mk(0, 2) * M_adj_Tk(0, 2);
+
+            if( det <= 1e-6 )
+            {
+                use_svd = true;
+                std::cout << "Oups, SVD serait utile!\n";
+                break;
+            }
+
+            if( det == 0.0 )
+            {
+                printf("Warning (polarDecomposition) : zero determinant encountered.\n");
+                break;
+            }
+
+            //mat_adj_T one-norm
+            float MadjT_one_norm = 0.0f;
+            for (int i = 0; i < 3; i++)
+            {
+                float col_abs_sum = fabs(M_adj_Tk(i, 0)) + fabs(M_adj_Tk(i, 1)) + fabs(M_adj_Tk(i, 2));
+                if (col_abs_sum > MadjT_one_norm)
+                    MadjT_one_norm = col_abs_sum;
+            }
+
+            //mat infinity-norm
+            float MadjT_inf_norm = 0.0;
+            for (int i = 0; i < 3; i++)
+            {
+                float row_sum = fabs(M_adj_Tk(i, 0)) + fabs(M_adj_Tk(i, 1)) + fabs(M_adj_Tk(i, 2));
+                if (row_sum > M_inf_norm)
+                    M_inf_norm = row_sum;
+            }
+
+            
+            float gamma = sqrtf(sqrtf((MadjT_one_norm * MadjT_inf_norm) / (M_one_norm * M_inf_norm * det * det)));
+            float g1 = gamma * 0.5f;
+            float g2 = 0.5f / (gamma * det);
+
+            for(int i = 0; i < 3; i++)
+            {
+                for(int j = 0; i < 3; j++)
+                    {
+                        Ek(i, j) = Mk(i, j);
+                        Mk(i, j) = g1 * Mk(i, j) + g2 * M_adj_Tk(i, j);
+                        Ek(i, j) -= Mk(i, j);
+                    }
+            }
+
+            //Ek one-norm
+            float E_one_norm = 0.0f;
+            for (int i = 0; i < 3; i++)
+            {
+                float col_abs_sum = fabs(Ek(i, 0)) + fabs(Ek(i, 1)) + fabs(Ek(i, 2));
+                if (col_abs_sum > E_one_norm)
+                    E_one_norm = col_abs_sum;
+            }
+
+            //M one-norm
+            float M_one_norm = 0.0f;
+            for (int i = 0; i < 3; i++)
+            {
+                float col_abs_sum = fabs(M(i, 0)) + fabs(M(i, 1)) + fabs(M(i, 2));
+                if (col_abs_sum > M_one_norm)
+                    M_one_norm = col_abs_sum;
+            }
+
+            //M infinity-norm
+            float M_inf_norm = 0.0;
+            for (int i = 0; i < 3; i++)
+            {
+                float row_sum = fabs(M(i, 0)) + fabs(M(i, 1)) + fabs(M(i, 2));
+                if (row_sum > M_inf_norm)
+                    M_inf_norm = row_sum;
+            }
+
+        } while ( E_one_norm > M_one_norm * tolerance );
+
+        
+        if(use_svd) // && force_rotation
+        {
+            // use the SVD algorithm to compute R
+            utils::Matrix3d Mm = M;
+            utils::Matrix3d Um, Vm;
+            utils::Vector3d lambda;
+            utils::Vector3d lambda_inverse;
+            int modified_SVD = 1;
```

Comment:
Does this happen ofter that the SVD is needed? Why not using it always? If it does not happen often, maybe a raise before is wiser, I feel the same for det==0.0... The reason is that I feel this happen if the provided matrix behaves badly.. And therefore return what should be the orthonormal matrix of a far-from-being-orthonormal-matrix is giving the rope to the user to hang themselves... But I may be wrong!

Reply:
I raised an error. But I am not sure if I agree with you since it might be encountered in the process of finding the right matrix. (not only user defined)

Reply:
Do you mean for any kind of matrices (not necessarily intended rotation matrix?)